### PR TITLE
Add prune integration tests for many edge cases

### PR DESCRIPTION
**Pull request by aawsome**
What is the purpose of this change? What does it change?
--------------------------------------------------------

Add prune integration tests for many edge cases like duplicate blobs, mixed blobs or all types of corrupt repo.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

This is part of #2718 that can be merged independently.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR
- I have not added documentation for the changes (in the manual)
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by aawsome** [comment]
Just realized that the "failing test" (it was a blob existing in the pack file but missing in the index) is in fact cured by "old" prune. So I changed this accordingly here.

File: cmd/restic/integration_test.go [hunk 1]
```diff
@@ -1283,6 +1283,103 @@ func TestPruneWithDamagedRepository(t *testing.T) {
...
+	datafile := filepath.Join("testdata", "repo-index-missing.tar.gz")
+	rtest.SetupTarTestFixture(t, env.base, datafile)
+
+	rtest.Assert(t, runCheck(CheckOptions{}, env.gopts, nil) != nil,
+		"check should have reported an error")
+	rtest.Assert(t, runPrune(env.gopts) != nil,
+		"prune should have reported an error")
+}
+
+func TestIndexMissingBlob(t *testing.T) {
```
Comment (MichaelEischer) [inline cmd/restic/integration_test.go:1299]: The other tests are also identical except for the tar-file, the check parameters and expected check result. Please merge these too.

Comment (MichaelEischer) [reply cmd/restic/integration_test.go:1299]: What does `IndexMissingBlob` check? That a single blob is not contained in the index, but still exists in the repository?

File: cmd/restic/integration_test.go [hunk 2]
```diff
@@ -1283,6 +1283,103 @@ func TestPruneWithDamagedRepository(t *testing.T) {
...
+
+func TestObsoleteIndex(t *testing.T) {
+	env, cleanup := withTestEnvironment(t)
+	defer cleanup()
+
+	datafile := filepath.Join("testdata", "repo-obsolete-index.tar.gz")
+	rtest.SetupTarTestFixture(t, env.base, datafile)
+
+	// check prints out an info but doesn't give an error for this setting.
+	testRunCheck(t, env.gopts)
```
Comment (MichaelEischer) [inline cmd/restic/integration_test.go:1365]: Can we adapt the test to check something?

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
How did you generate the test data sets? I wonder if there's an easy way to generate those on the fly?

**Comment by aawsome** [comment]
> How did you generate the test data sets? I wonder if there's an easy way to generate those on the fly?

Honestly, by playing a lot around with creating backups, deleting manually single files from the repo, running `prune` and `rebuild-index` and sometimes even changing the source code to get the desired test repository.
I don't feel like I'm able to code that with small effort.

Also I think that keeping the repos as tar files "hardens" the tests cases in the sense that future code changes will not spoil the test cases. 
E.g. the code-change that removed mixed pack creation effectively prevents to generate those test settings on the fly... Similar things could happen with the other cases in future.

File: cmd/restic/integration_test.go (hunk 1 above)
Comment (aawsome) [reply cmd/restic/integration_test.go:1299]: Yes - `repo-index-missing-blob.tar.gz` is a repo where the packs are fine but a single (used) blob is missing from the index. With current `prune` this will be simply fixed (as `prune` always starts with the `rebuild-index` step), but after #2718 this will result in a prune error, which has to be explicitly fixed, e.g. by `rebuild-index`.

File: cmd/restic/integration_test.go [hunk 3]
```diff
@@ -1283,6 +1283,103 @@ func TestPruneWithDamagedRepository(t *testing.T) {
 	t.Log(err)
 }
 
+func TestIndexMissing(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (aawsome) [reply cmd/restic/integration_test.go:1286]: I now use just two functions and also added comments about what is tested.

File: cmd/restic/integration_test.go (hunk 2 above)
Comment (aawsome) [reply cmd/restic/integration_test.go:1365]: It actually tests that restic `check` and `prune` still work with an obsolete index... :wink:

File: cmd/restic/integration_test.go (hunk 1 above)
Comment (aawsome) [reply cmd/restic/integration_test.go:1299]: The tests are now all merged.

File: cmd/restic/integration_test.go [hunk 4]
```diff
@@ -1283,6 +1283,67 @@ func TestPruneWithDamagedRepository(t *testing.T) {
 	t.Log(err)
 }
 
+// Test repos for edge cases
+func TestEdgeCaseRepos(t *testing.T) {
+	opts := CheckOptions{}
+
+	// repo where index is completely missing
+	// => check and prune should fail
+	testEdgeCaseRepo(t, "repo-index-missing.tar.gz", opts, false, false)
```
Comment (MichaelEischer) [inline cmd/restic/integration_test.go:1292]: Please use `t.Run(...)` to make these subtests selectable. That could come in handy when hunting a specific problem.

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
> Also I think that keeping the repos as tar files "hardens" the tests cases in the sense that future code changes will not spoil the test cases.

That will also in parts verify that we don't break repository backwards compatibility.

I'd like to have the edge case tests as subtests to allow selecting a specific one. Besides that the code is ok. (The travis build error will disappear on the next CI run.)

File: cmd/restic/integration_test.go (hunk 4 above)
Comment (aawsome) [reply cmd/restic/integration_test.go]: Thanks for the hint about how to make subtests - I wasn't aware of this.
Is changed now.

**Review by MichaelEischer — APPROVED** [review]
LGTM. Thanks :-)
